pkg/command/apitoken_test.go

```diff
@@ -0,0 +1,204 @@
+				GlobalOpts: &configtypes.GlobalServer{
+					Auth: configtypes.GlobalServerAuth{
+						Issuer: "https://example.com",
+					},
+				},
+				AdditionalMetadata: map[string]interface{}{
+					config.TanzuIdpTypeKey: config.UAAIdpType,
+				},
+			},
+			tanzuLoginErr: nil,
+			wantErr:       false,
+			output: `==
+
+API Token Generation Successful! Your generated API token is: refresh-token
+
+Token Expiration: The token will expire on 2024-01-01 05:00:00.
+Please copy and save your token securely. Note that you will need to regenerate a new token before expiration time and login again to continue using the CLI.
+
+For non-interactive login use the API token as follow: TANZU_API_TOKEN=<token> tanzu login --endpoint <tanzu-platform-endpoint>
+`,
+		},
+		{
+			name:          "no active context",
+			context:       nil,
+			tanzuLoginErr: nil,
+			wantErr:       true,
+			errMsg:        "No active context of type `tanzu`. Please login to Tanzu Platform first to generate the API token",
+		},
+		{
+			name: "invalid active context",
+			context: &configtypes.Context{
+				Name:        "fakecontext",
+				ContextType: configtypes.ContextTypeTanzu,
+				AdditionalMetadata: map[string]interface{}{
+					config.TanzuIdpTypeKey: config.UAAIdpType,
+				},
+			},
+			tanzuLoginErr: nil,
+			wantErr:       true,
+			errMsg:        "Invalid active context of type `tanzu`. Please login to Tanzu Platform first to generate the API token",
+		},
+		{
+			name: "invalid IDP type",
+			context: &configtypes.Context{
+				Name:        "fakecontext",
+				ContextType: configtypes.ContextTypeTanzu,
+				GlobalOpts: &configtypes.GlobalServer{
+					Auth: configtypes.GlobalServerAuth{
+						Issuer: "https://example.com",
+					},
+				},
+				AdditionalMetadata: map[string]interface{}{
+					config.TanzuIdpTypeKey: "other",
+				},
+			},
+			tanzuLoginErr: nil,
+			wantErr:       true,
+			errMsg:        "invalid IDP type for the active context. Only UAA IDP type is supported for generating API token",
+		},
+		{
+			name: "TanzuLogin error",
+			context: &configtypes.Context{
+				Name:        "fakecontext",
+				ContextType: configtypes.ContextTypeTanzu,
+				GlobalOpts: &configtypes.GlobalServer{
+					Auth: configtypes.GlobalServerAuth{
+						Issuer: "https://example.com",
+					},
+				},
+				AdditionalMetadata: map[string]interface{}{
+					config.TanzuIdpTypeKey: config.UAAIdpType,
+				},
+			},
+			tanzuLoginErr: errors.New("TanzuLogin error"),
+			wantErr:       true,
+			errMsg:        "unable to login, TanzuLogin error",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			setupEnv()
+			defer teardownEnv()
+
+			cmd := &cobra.Command{}
+			outputBuf := &bytes.Buffer{}
+			cmd.SetOutput(outputBuf)
+
+			mockTimeService := &mockTimeService{}
+			timeutils.Now = mockTimeService.Now
+			defer func() { timeutils.Now = time.Now }()
+
+			mockTanzuLogin := &mockTanzuLogin{}
+			uaa.TanzuLogin = mockTanzuLogin.TanzuLogin
```

Comment:
Should we remember the uaa.TanzuLogin and restore it on a defer so we minimize the impact of the function patching?